gh-131253: free-threaded build support for pystats

nascheme opened the pull request:
Allow the `--enable-pystats` build option to be used with free-threading.  For the free-threaded builds, the stats structure is allocated per-thread and then periodically merged into a per-interpter stats structure (on thread exit or when the reporting function is called).

Summary of changes:

* Add `pystats` to the PyThreadState structure.  This is a pointer to the PyStats structure if recording of stats is enabled.  This is used for both the free-threaded and default builds.

* For the free-threaded build, the _PyThreadStateImpl structure gains a pystats_struct member.  This is allocated per-thread when recording of pystats is first enabled.

* replace `_Py_stats` references with `_PyStats_GET()`

* move pystats logic from Python/specialize.c into Python/pystats.c

* move the pystats global state into the interpreter structure

* add some free-threaded specific stat counters

Notes and potential issues:

* The FTStats counts will need review.  I wasn't sure about the naming or even how useful these might be.  For `mutex_sleeps`, we need to determine what is the most useful thing to measure.  I increment that count whenever we have to "spin" on a mutex.

* The verbose code for `print_*` and `merge_*` is perhaps not ideal.  I considered making this data driven instead.  However, I think that actually would add complexity, not reduce it.

* I plan on adding two levels of pystats recording: default and extended.  The "default" level would be enabled by default (or at least could be) and would only include stats that are relatively cheap to record.  The "extended" level would be like what the current `--enable-pystats` does (counting things like INCREF/DECREF, which is expensive).


<!-- gh-issue-number: gh-131253 -->
* Issue: gh-131253
<!-- /gh-issue-number -->

nascheme committed:
gh-131253: free-threaded build support for pystats

Allow the --enable-pystats build option to be used with
free-threading.  For the free-threaded builds, the stats structure is
allocated per-thread and then periodically merged into a global stats
structure (on thread exit or when the reporting function is called).

Summary of changes:

* introduce _Py_tss_stats thread-local variable.  This is set
  when stats are on, replacing the _Py_stats global that's used
  in the non-free-threaded build.

* replace _Py_stats references with _PyStats_GET()

* move pystats logic from Python/specialize.c into Python/pystats.c

* add some free-threaded specific stat counters

nascheme committed:
Make pystats global state be per-interpreter.

nascheme committed:
Add NEWS.

nascheme committed:
Merge 'origin/main' into gh-131253-pystats-ft

colesbury commented on Include/cpython/pystats.h:
```diff
@@ -173,22 +182,57 @@ typedef struct _stats {
+#define _Py_STATS_COND_EXPR(cond, expr) \
+    do { \
+        PyStats *s = _PyStats_GET(); \
+        if (s != NULL && cond) { \
```
```suggestion
        if (s != NULL && (cond)) { \
```

colesbury reviewed (commented):
The management of the thread local variable seems complex. Can we keep the thread-local stats on `PyThreadState` and have `_PyStats_GET()` return something like `_PyThreadState_GET()->stats`? How much would the extra indirection slow down the stats build?

nascheme committed:
Fix some data race issues.

Need to do a merge before reporting (I lost that bit of code on a
re-factor).

Fix various issues with data races.  When merging from all threads, we
need to stop-the-world to avoid races.  When toggling on or off, also
need to stop-the-world.  Remove the need for locking for
_PyStats_Attach().

nascheme committed:
Remove _Py_tss_stats thread local.

Add pystats pointer to PyThreadState and use from there instead.  This
is slightly slower but shouldn't matter in practice.  This simplifies
the attach/detach logic as well.

nascheme committed:
Merge 'origin/main' into gh-131253-pystats-ft

nascheme committed:
Improve _Py_STATS_COND_EXPR macro.

nascheme committed:
Improve comment for ts->pystats member.

kumaraditya303 commented on Include/cpython/pystats.h:
```diff
@@ -109,6 +109,18 @@ typedef struct _gc_stats {
+    // number of times that the QSBR mechanism polled (compute read sequence value)
+    uint64_t qsbr_polls;
+    // number of times stop-the-world mechanism was used
+    uint64_t world_stops;
```
It would be useful to have the total duration of stop-the-world pause too

markshannon replied:
In general, the stats should count things and be independent of timing.
With stats enabled timings will be so different from normal, as to make the timings useless anyway.

markshannon commented on Objects/object.c:
```diff
@@ -2418,6 +2418,13 @@ _PyObject_InitState(PyInterpreterState *interp)
         return _PyStatus_NO_MEMORY();
     }
+#endif
+#ifdef Py_STATS
```
This has been moved from initconfig.c?
Is initconfig called too late?

nascheme replied:
It's been a while since I worked on this but, if I recall, the problem is that initconfig doesn't give you the interpreter.  The stats enabled flag is now per-interpreter whereas before it was a process global (the value of the `_Py_stats` pointer being non-NULL).

sergey-miryanov commented on Python/pystats.c:
Should we use atomic load here?

nascheme replied:
I think this is safe without an atomic.  `interp->pystats_struct` only gets assigned in two places.  First, when an interpreter is first created, before any threads are running.  Second, in `stats_toggle_on_off()`, where we hold the stats lock.  We hold the lock here and so we ensure memory ordering.

sergey-miryanov replied:
Thanks for explanation!

sergey-miryanov commented on Python/pystats.c:
Do we rewrite the pystats_struct for the interp each time when we create a new thread if the pystats_enabled is true?

nascheme replied:
Oh, this is a bug.  Thank you for finding it!  I restructured how pystats_struct is initialized, fixing this.  I think the new code is slightly better organized as well.

sergey-miryanov replied:
Thanks for fixing!

sergey-miryanov commented on Python/pystats.c:
It seems that we perform a lot of unnecessary work in `stats_merge_thread` if `zero` is true. Is this intentional?

nascheme replied:
The operation is to add the counts from the thread local stats to the per-interpreter stats.  Once that's done, we optionally zero the thread local stats.  I don't think there is unnecessary work as we want to merge those counts in before we zero them.

sergey-miryanov replied:
IIUC, we copy thread local stats from `ts` to `interp->pystats_struct`, zeroize `ts` _unconditionally_, then if `zero` if true we zeroize `interp->pystats_struct`. IIUC, we may just zeroize `ts` if `zero` is true.

nascheme replied:
I created separate "zero" functions rather than having the "zero" boolean option.  I think it's more clear and it's a bit more efficient.

sergey-miryanov replied:
Thanks! It looks much clearer to understanding!

nascheme committed:
Cleanup init of interp->pystats_struct.

Add the _PyStats_InterpInit() function which will set pystats_enabled
flag and allocate pystats_struct as required.  We shouldn't be putting
logic in _PyObject_InitState().  This also fixes a bug where
interp->pystats_struct was allocated repeated rather than once per
interpreter.

nascheme committed:
Split stats zeroing into separate functions.

This is a bit more efficient, with the downside of a bit extra code.
For _PyStats_Clear(), we don't need to do the merging operation.  I
think the code is a bit more clear this way.

nascheme committed:
Merge 'origin/main' into gh-131253-pystats-ft

sergey-miryanov commented on Python/pystats.c:
IIUC, we use `PyMem_RawCalloc` to allocate new interp memory. Then the default value of `interp->pystats_enabled` is 0, so if `interp->config._pystats` is false, we don't need to initialize it.

sergey-miryanov reviewed (approved):
LGTM

kumaraditya303 commented on Include/cpython/pystate.h:
```diff
@@ -239,6 +248,21 @@ PyAPI_FUNC(void) PyThreadState_EnterTracing(PyThreadState *tstate);
+static inline PyStats*
+_PyThreadState_GetStatsFast(void)
+{
+    if (_Py_tss_tstate == NULL) {
```
I think you should be able to use `_PyThreadState_GetCurrent` instead of accessing here

nascheme replied:
I don't want `_PyThreadState_GetCurrent()` because I want to inline the access to `_Py_tss_tstate->pystats`.  The `_PyStats_GET()` function is used a lot from hot paths.  

I could potentially use `_PyThreadState_GET()`.  However, I don't see an easy way to make that work.  The header that defines that inline function is not always included, e.g. for Programs/_freeze_module.  Accessing `_Py_tss_state` directly is the simplest way I could find to do it.

kumaraditya303 commented on Python/pystats.c:
Since this loop happens when the world is paused, all the other threads will have _active=0, why is this check needed?

nascheme replied:
Yup, you are correct.  This "active" check shouldn't be here.  I fixed this and a couple other toggle on/off issues I found.  I also added a unit test since obviously I didn't test this toggle functionality well enough since it wasn't working in other threads.

nascheme committed:
Fix a couple bugs related to stats allocation.

Need to allocate in stats_toggle_on_off() if the thread never had
stats enabled yet.  For zeroing, handle case that struct is not
allocated.  Remove bogus ts->_status.active check.

nascheme committed:
Add unit test for pystats.

nascheme committed:
Merge 'origin/main' into gh-131253-pystats-ft

kumaraditya303 commented on Lib/test/test_pystats.py:
```diff
@@ -0,0 +1,215 @@
+import unittest
+from test.support import script_helper
+
+# This function is availible for the --enable-pystats config.
```
```suggestion
# This function is available for the --enable-pystats config.
```

nascheme committed:
Typo fix for Lib/test/test_pystats.py

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

kumaraditya303 commented on Include/cpython/pystate.h:
```diff
@@ -217,6 +217,15 @@ struct _ts {
     _PyRemoteDebuggerSupport remote_debugger_support;
+
+#ifdef Py_STATS
+    // Pointer PyStats structure, NULL if recording is off.  For the
```
```suggestion
    // Pointer to PyStats structure, NULL if recording is off.  For the
```

kumaraditya303 commented on Include/cpython/pystate.h:
```diff
@@ -239,6 +248,21 @@ PyAPI_FUNC(void) PyThreadState_EnterTracing(PyThreadState *tstate);
 PyAPI_FUNC(void) PyThreadState_LeaveTracing(PyThreadState *tstate);
 
+#ifdef Py_STATS
+#if defined(HAVE_THREAD_LOCAL) && !defined(Py_BUILD_CORE_MODULE)
```
With https://github.com/python/cpython/pull/140690, `HAVE_THREAD_LOCAL` check is not needed

nascheme replied:
I think it's still needed.  In pycore_pystate.h we have:
```
#if defined(HAVE_THREAD_LOCAL) && !defined(Py_BUILD_CORE_MODULE)
extern _Py_thread_local PyThreadState *_Py_tss_tstate;
#endif
```
I could change the code to:
```
#if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
```
But I don't think that's better.

kumaraditya303 commented on Include/cpython/pystate.h:
```diff
@@ -239,6 +248,21 @@ PyAPI_FUNC(void) PyThreadState_EnterTracing(PyThreadState *tstate);
 
+#ifdef Py_STATS
+#if defined(HAVE_THREAD_LOCAL) && !defined(Py_BUILD_CORE_MODULE)
+extern _Py_thread_local PyThreadState* _Py_tss_tstate;
```
```suggestion
extern _Py_thread_local PyThreadState *_Py_tss_tstate;
```

kumaraditya303 commented on Include/cpython/pystate.h:
```diff
@@ -239,6 +248,21 @@ PyAPI_FUNC(void) PyThreadState_EnterTracing(PyThreadState *tstate);
+#if defined(HAVE_THREAD_LOCAL) && !defined(Py_BUILD_CORE_MODULE)
+extern _Py_thread_local PyThreadState* _Py_tss_tstate;
+
+static inline PyStats*
```
```suggestion
static inline PyStats *
```

kumaraditya303 commented on Include/cpython/pystats.h:
```diff
@@ -173,22 +185,48 @@ typedef struct _stats {
+// Export for most shared extensions
+PyAPI_FUNC(PyStats *) _PyStats_GetLocal(void);
+
+#if defined(HAVE_THREAD_LOCAL) && !defined(Py_BUILD_CORE_MODULE)
```
Same

nascheme committed:
Typo fix for comment in Include/cpython/pystate.h

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

nascheme committed:
Code style tweaks.
